### PR TITLE
Data Model Changes for Tags

### PR DESCRIPTION
**Pull request by tap2k**
Passes all the tests!

### CODE REVIEW

File: apps/settings.py [hunk 1]
```diff
@@ -198,7 +198,7 @@
     'rest_framework',
     'corsheaders',
     'social.apps.django_app.default',
-    'swampdragon'
+    'swampdragon',
```
Comment (vanwars) [inline apps/settings.py]: let's keep swampdragon lib out of settings.py

File: apps/site/api/fields/list_field.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    def get_value(self, dictionary):
+        if self.field_name not in dictionary:
+            if getattr(self.root, 'partial', False):
+                return empty
+        # We override the default field access in order to support
+        # lists in HTML forms.
+        if html.is_html_input(dictionary):
+            val = dictionary.get(self.field_name, None)
+            if val:
+                print "VALUE IS:" + val
```
Comment (vanwars) [inline apps/site/api/fields/list_field.py]: nitpick: drop print statement

File: apps/site/api/serializers/base_serializer.py [hunk 3]
```diff
@@ -57,8 +58,17 @@ class Meta:
 
 
 class BaseNamedSerializer(BaseSerializer):
-    tags = serializers.CharField(required=False, allow_null=True, label='tags',
-                                    help_text='Tag your object here', allow_blank=True)
+    #tags = serializers.ListField(child=serializers.CharField(), required=False, allow_null=True, label='tags', help_text='Tag your object here')
+    #tags = serializers.ListField(child=serializers.CharField(), required=False, allow_null=True, label='tags', help_text='Tag your object here')
+    #tags = StringListField(required=False, allow_null=True, label='tags', help_text='Tag your object here')
```
Comment (vanwars) [inline apps/site/api/serializers/base_serializer.py]: Can we get rid of these?

File: apps/site/api/serializers/base_serializer.py [hunk 4]
```diff
@@ -80,6 +90,7 @@ def get_projects(self):
             return models.Project.objects.all()
 
     class Meta:
+        model = BaseNamed
```
Comment (vanwars) [inline apps/site/api/serializers/base_serializer.py]: Not sure we need this...
model = BaseNamed

File: apps/site/api/serializers/print_serializer.py [hunk 5]
```diff
@@ -39,8 +38,9 @@ def get_fields(self, *args, **kwargs):
         source='name',
         required=False,
         allow_blank=True)
-    tags = serializers.CharField(required=False,
-        allow_blank=True, help_text='Tag your object here')
+    #tags = serializers.CharField(required=False,
+    #    allow_blank=True, help_text='Tag your object here')
+    tags = serializers.ListField(child=serializers.CharField(),required=False, allow_null=True, label='tags', help_text='Tag your object here')
```
Comment (vanwars) [inline apps/site/api/serializers/print_serializer.py]: can we remove these comments?

File: apps/site/api/tests/presentation_tests.py [hunk 6]
```diff
@@ -10,12 +10,12 @@
 class ApiPresentationTest(object):
     name = 'New Presentation'
     description = 'description of presentation'
-    tags = "a, b, c"
+    tags = [u'a',u'b',u'c']
```
Comment (vanwars) [inline apps/site/api/tests/presentation_tests.py]: I'm really confused by this. Why for some tests are there arrays and others strings? I think this test needs to be developed.

File: apps/site/api/tests/print_tests.py [hunk 7]
```diff
@@ -139,16 +139,16 @@ def setUp(self):
             'zoom': {'read_only': True, 'required': False, 'type': 'integer'},    
         })
 
-    def test_update_print_using_patch(self, **kwargs):
-        name, description, tags = 'A', 'B', 'C'
+    def test_tnt_using_patch(self, **kwargs):
+        name, description, tags = 'A', 'B', [u'C']
```
Comment (vanwars) [inline apps/site/api/tests/print_tests.py]: does this work?

**Comment by vanwars** [comment]
Hi Tap,

I'm getting some errors, and I'm also confused about the data type that a "tags" field should accept. I was expecting the field to accept a comma-delimited string for POST / PUT requests, and return a list for GET requests. However, when I tested POST / PUT with photos, it worked using the /api/0/photos/ browser-based viewer for the following "a, b, c", but threw the following bad request error when doing the same post using the map editor (/maps/edit/new/):

"{"tags":["Expected a list of items but got type \"unicode\"."]}"

Not sure what's going on there, but maybe you could take a look. Also, it threw an error when I didn't include a "tags" value in the POST / PUT. I think you need to write some additional tests that check that requests accept null or empty values. Lastly, some of your tests confused me because sometimes they're testing for strings ("a, b, c") and sometimes for lists ([u'C']).

File: apps/site/api/tests/photo_tests.py [hunk 8]
```diff
@@ -143,6 +162,7 @@ def test_update_photo_using_put(self, **kwargs):
         self.assertEqual(updated_photo.geometry.y, point['coordinates'][1])
         self.assertEqual(updated_photo.geometry.x, point['coordinates'][0])
         self.assertEqual(updated_photo.extras, extras)
+        self.assertEqual(updated_photo.tags, tags)
```
Comment (vanwars) [inline apps/site/api/tests/photo_tests.py]: How is this working?

File: apps/site/api/tests/photo_tests.py [hunk 9]
```diff
@@ -125,16 +129,31 @@ def tearDown(self):
...
-                                'extras': json.dumps(extras)
-                            }),
-                            HTTP_X_CSRFTOKEN=self.csrf_token,
-                            content_type="application/x-www-form-urlencoded"
-                        )
+                    data=json.dumps({
+                        'geometry': json.dumps(point),
+                        'name': name,
+                        'caption': description,
+                        'tags' : tags,
```
Comment (vanwars) [inline apps/site/api/tests/photo_tests.py]: How is this working if its an array?

File: apps/site/api/tests/project_tests.py [hunk 10]
```diff
@@ -34,17 +34,17 @@ def tearDown(self):
     def test_create_project_using_post(self, **kwargs):
         name = 'New Project!'
         description = 'New project description'
-        tags = "d, e, f"
+        tags = [u'd', u'e', u'f']
```
Comment (vanwars) [inline apps/site/api/tests/project_tests.py]: array

File: apps/site/api/views/tags_views.py [hunk 11]
```diff
@@ -1,60 +1,10 @@
...
-from tagging.models import Tag, TaggedItem
-from localground.apps.site import models
-from localground.apps.lib.helpers.generic import FastPaginator
-from rest_framework.permissions import AllowAny
-from localground.apps.site.api.views import *
+from rest_framework import status
+from rest_framework.views import APIView
+from rest_framework.response import Response
 
+class TagList(APIView):
```
Comment (vanwars) [inline apps/site/api/views/tags_views.py]: This needs tests

File: apps/site/api/tests/layer_tests.py [hunk 12]
```diff
@@ -10,12 +10,12 @@
 class ApiLayerTest(object):
     name = 'New Layer Name'
     description = 'Test layer description'
-    tags = 'a, b, c'
+    tags = [u'a',u'b',u'c']
```
Comment (vanwars) [inline apps/site/api/tests/layer_tests.py]: please switch this to 'a, b, c' for readability.

File: apps/site/api/tests/snapshot_tests.py [hunk 13]
```diff
@@ -27,7 +27,7 @@ def get_metadata():
 class ApiSnapshotTest(object):
     name = 'New Snapshot Name'
     description = 'Test description'
-    tags = 'a, b, c'
+    tags = ['a','b','c']
```
Comment (vanwars) [inline apps/site/api/tests/snapshot_tests.py]: ditto list -> string

File: apps/site/api/views/tags_views.py [hunk 14]
```diff
@@ -1,60 +1,16 @@
...
-                )
-            return Tag.objects.usage_for_queryset(queryset, counts=True)
-            # return Tag.objects.cloud_for_model(model_class)
+from rest_framework import status
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from localground.apps.site.models import Audio, Photo, Print, Video
+from django.db.models import Func, F
+
+class TagList(APIView):
```
Comment (vanwars) [inline apps/site/api/views/tags_views.py]: Needs a test

**Comment by vanwars** [comment]
* Minor comments above
* The tagging UI widget in http://localhost:7777/profile/photos/ is really nice. For whatever reason, it now display an error message when I update, but since Keshav's migrating this to backbone, it's fine for now. Also, I. I can take a stab at migrating this to Map Backbone UI.
* Tags endpoint still needs a test

File: apps/site/api/views/tags_views.py (hunk 14 above)
Comment (tap2k) [inline apps/site/api/views/tags_views.py]: Ill need to think about the test for this as it is a property of the global state of the app. Anyway can we punt on this until I finish the endpoint?

Comment (vanwars) [inline apps/site/api/views/tags_views.py]: I'd prefer not to punt on tests because they don't take long to make and
they're really helpful to flag regressions. I didn't realize this endpoint
wasn't done. If you're still working on this PR and don't want to merge it
yet, let me know, but I'd prefer not to roll new functionality into master
w/o tests. You could also just start by making a stub that (a) inserts one
tagged object into the database for each data type (photo, audio, and
map_image), (b) queries the endpoint, and (c) ensures that there are at
least n tags returned in the array. It's not exhaustive, but it will still
be a really useful indicator to know whether or not autocomplete is
up-and-running, or down.

Comment (tap2k) [inline apps/site/api/views/tags_views.py]: OK. Its done-ish but I was going to add the ability to query by project. When I retrieve objects in Django, is there way to make sure I only retrieve those for which the user has permissions? Im worried about Tags leaking.

File: apps/site/api/views/tags_views.py [hunk 15]
```diff
@@ -1,60 +1,16 @@
...
+
+  def get(self, request, *args, **kw):
+    phototags = list(Audio.objects.annotate(arr_els=Func(F('tags'),function='unnest')).values_list('arr_els', flat=True).distinct())
```
Comment (vanwars) [inline apps/site/api/views/tags_views.py]: Here, to restrict scope to only authorized projects, I think you can do this:

list(Audio.objects.get_objects(request.user).annotate(arr_els=Func(F('tags'),function='unnest')).values_list('arr_els', flat=True).distinct()). Ditto for others.

Comment (tap2k) [inline apps/site/api/views/tags_views.py]: Are those objects owned by that user or accessible by him/her?

Comment (vanwars) [inline apps/site/api/views/tags_views.py]: Either owned by or shared with the user (under the hood, uses a bunch of
views, created here: /localground/apps/sql/custom.sql)

**Comment by vanwars** [comment]
Looks great, Tap! I also added an additional permissions test (mostly because I was curious), to see if the api/0/tags endpoint honored permissions. It does! Will merge this as soon as the Travis tests run.
